### DISCUSSION TITLE
Potential Causes of Memory Leaks

### DISCUSSION THREAD

**Discussion by hwhitehead** [PrincetonUniversity/athena#558] [Q&A] [answered]
I have a simulation I am running that appears to use more memory as the simulation progresses. While the simulation does use AMR, I have checked and the number of MeshBlocks stays the same so memory usage should not be increasing with time.

I wanted to check to make sure I am following Athena++ best practices in case I have introduced a memory leak into the system somewhere. While the custom pgen for this simulation is quite extensive, I have tried to keep to the general Athena++ formalism. 

I have not attempted to make any allocations directly to heap; all variables exist on the stack or are generated via the NewAthenaArray constructor. If I create an array using this method, am I correct in saying it should automatically be deallocated when it leaves scope, or do I need to manually free this memory?

A code snippet for example:
```
void Mesh::UserWorkInLoop() {
  # create some new array of 10 Reals
  AthenaArray<Real> foo;
  foo.NewAthenaArray(10);

  # do some calculation using foo

} # foo and associated memory freed automatically here?
```

From a quick scan of the AthenaArray [definition](https://github.com/PrincetonUniversity/athena/blob/e58b549877582762bae52ec9bc91b5bd0de43ee3/src/athena_arrays.hpp#L26) I think ```foo``` should be freed automatically, but if this is not the case then this may be a potential issue in my system.

If there are any other obvious potential memory leak sources specific to Athena++ that I should be aware of please let me know.

**Comment by felker** [comment]
It is deallocated automatically when it leaves scope here

**Comment by felker** [comment]
How significant is the growth in memory?

**Reply by hwhitehead** [reply]
The growth IS significant, over ~10,000 cycles the memory consumption approximately doubles.

**Reply by hwhitehead** [reply]
Some more information, memory usage at launch is around 10GB (which is reasonable given the number of cells, registers etc) but after a few hours of running the memory consumption has grown to 20GB+. This is using an MPI parallelized system of 25 cores each across two nodes. Memory growth is also observed in a single node system for a variety of core counts.

**Comment by c-white** [comment]
Does this happen with a single MPI rank and single thread? If it does, it might be worth running the executable through `valgrind` to see if it gives any useful feedback (though it might flag things unrelated to your issue). If it doesn't, the first place to look would be any additional communication you may have written into the code.

**Reply by hwhitehead** [reply]
For runs in parallel, the growth is observed on all threads (though perhaps not perfectly evenly). I am waiting on results to see if memory growth occurs for unparallelised runs. 

I am also running some tests with valgrind and varying levels of parallelization to see if I can pin down the root.

Cheers for the guidance!

**Reply by hwhitehead** [reply]
Runs in single do not seem to experience the memory growth observed in parallel, suggesting the error originates somewhere in my cross-rank communication. I have generally stuck to the format suggested in [this](https://github.com/PrincetonUniversity/athena/discussions/464#discussioncomment-4117711) previous discussion, but will take a closer look to see what mistakes I may have made.

**Comment by hwhitehead** [comment]
Memory gain is present for standard test simulations (e.g. orszag-tang) on a clean, updated install of Athena++. I expect this means I have some issues with my compiler/MPI libraries, I will contact my cluster admin to see if I can close this issue.

**Reply by hwhitehead** [reply, marked as answer]
Issue is now resolved. It appears there was some issue with how the cluster libraries were compiled, rather than any specific interaction with Athena++. Marking this issue as answered, thanks all for the advice.
